kayenta-web/src/main/java/com/netflix/kayenta/controllers/MetricSourcesController.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2017 Google, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.kayenta.controllers;
+
+import com.netflix.kayenta.metrics.MetricsService;
+import com.netflix.kayenta.metrics.MetricsServiceRepository;
+import com.netflix.kayenta.security.AccountCredentials;
+import com.netflix.kayenta.security.AccountCredentialsRepository;
+import com.netflix.kayenta.security.CredentialsHelper;
+import io.swagger.annotations.ApiOperation;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.netflix.kayenta.security.AccountCredentials.Type.METRICS_STORE;
+
+@RestController
+@RequestMapping("/metricSources")
+public class MetricSourcesController {
+
+  private final AccountCredentialsRepository accountCredentialsRepository;
+  private final MetricsServiceRepository metricsServiceRepository;
+
+  @Autowired
+  public MetricSourcesController(AccountCredentialsRepository accountCredentialsRepository, MetricsServiceRepository metricsServiceRepository) {
+    this.accountCredentialsRepository = accountCredentialsRepository;
+    this.metricsServiceRepository = metricsServiceRepository;
+  }
+
+  @ApiOperation(value = "Retrieve a list of all configured metrics sources")
+  @RequestMapping(method = RequestMethod.GET)
+  List<MetricSourceDetail> list() {
+    Set<AccountCredentials> metricAccountCredentials = CredentialsHelper.getAllAccountsOfType(METRICS_STORE, accountCredentialsRepository);
+
+    return metricAccountCredentials
+      .stream()
+      .map(account -> {
+        MetricsService metricsService =
+          metricsServiceRepository
+            .getOne(account.getName())
+            .orElseThrow(() -> new IllegalArgumentException("No metrics service was configured; unable to read from metrics store."));
+
+        return MetricSourceDetail
+          .builder()
+          .name(account.getName())
+          .type(metricsService.getType())
+          .locations(metricsService.getLocations(account.getName()))
```

Comment:
Since the logic in `metricsService.getLocations` and `metricsService. getRecommendedLocations` operate only on the credentials objects, I can't see why this needs to go through the metrics service at all? Would all be much simpler if the `locations` and `recommendedLocations` properties were only exported from the credentials class, and not from the metrics service class.

Reply:
Naming aside, the `CredentialsController` already returns all configured services. Why not just add a `filter` to that to specify what types of accounts you want to retrieve, and then pass `METRICS_STORE` or something similar to it?

It will return anything defined (and not hidden) on the credentials classes, so you would get your `locations` and `recommendedLocations` for free, no?

Those are the 2 new bits we're trying to retrieve, the name & type are already being retrieved.

Reply:
I started adding this to the credentials class, but I think the location is specific to a metric source, not to the generic credential. It didn't feel right to add this to the credentials, so went with the more specific metric service.

Reply:
That's a fair point. Just thought it would be simpler without adding the new service which just ends up back at the credentials anyway.

I can live with it either way.